src/Libraries/Microsoft.Extensions.AI.Abstractions/Utilities/DataContentAsyncEnumerableStream.cs

```diff
@@ -0,0 +1,228 @@
+                    break;
+                }
+
+                if (!_enumerator.Current.Data.HasValue)
+                {
+                    _isCompleted = true;
+                    break;
+                }
+
+                _remainingData = _enumerator.Current.Data.Value.ToArray();
+                _remainingDataOffset = 0;
+            }
+        }
+
+        return bytesRead;
+    }
+#endif
+
+#if NET8_0_OR_GREATER
+    /// <inheritdoc/>
+    public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
+    {
+        if (_isCompleted)
+        {
+            return 0;
+        }
+
+        int bytesRead = 0;
+        int totalToRead = buffer.Length;
+
+        while (bytesRead < totalToRead)
+        {
+            // If there's still data in the current chunk
+            if (_remainingDataOffset < _remainingData.Length)
+            {
+                int bytesToCopy = Math.Min(totalToRead - bytesRead, _remainingData.Length - _remainingDataOffset);
+                _remainingData.AsSpan(_remainingDataOffset, bytesToCopy)
+                              .CopyTo(buffer.Span.Slice(bytesRead, bytesToCopy));
+
+                _remainingDataOffset += bytesToCopy;
+                bytesRead += bytesToCopy;
+                _position += bytesToCopy;
+            }
+            else
+            {
+                // If the first chunk was never read, attempt to read it now
+                if (_position == 0 && _firstDataContent is not null && _firstDataContent.Data.HasValue)
+                {
+                    _remainingData = _firstDataContent.Data.Value.ToArray();
+                    _remainingDataOffset = 0;
+                    continue;
+                }
+
+                // Move to the next chunk in the async enumerator
+                if (!await _enumerator.MoveNextAsync().ConfigureAwait(false) ||
+                    !_enumerator.Current.Data.HasValue)
+                {
+                    _isCompleted = true;
+                    break;
+                }
+
+                _remainingData = _enumerator.Current.Data.Value.ToArray();
+                _remainingDataOffset = 0;
+            }
+        }
+
+        return bytesRead;
+    }
+#endif
+
+#if NET8_0_OR_GREATER
+    /// <inheritdoc/>
+    public override async ValueTask DisposeAsync()
+    {
+        await _enumerator.DisposeAsync().ConfigureAwait(false);
+
+        await base.DisposeAsync().ConfigureAwait(false);
+    }
+#else
+    public async ValueTask DisposeAsync()
+    {
+        await _enumerator.DisposeAsync().ConfigureAwait(false);
+
+        Dispose();
+    }
+#endif
+
+    /// <inheritdoc/>
+    public override void Write(byte[] buffer, int offset, int count)
+    {
+        throw new NotSupportedException();
+    }
+
+    /// <inheritdoc/>
+    protected override void Dispose(bool disposing)
+    {
+        if (disposing)
+        {
+            var task = Task.Run(_enumerator.DisposeAsync);
```

Comment:
This needs to wait for disposal to complete.

Reply:
Improved the logic a bit